src/Native/System.Native/pal_networking.cpp

```diff
@@ -392,3 +401,84 @@ extern "C" int32_t GetHostName(uint8_t* name, int32_t nameLength)
+    {
+        assert(false && "getifaddrs failed.");
+    }
+
+    for (current = headAddr; current != nullptr; current = current->ifa_next)
+    {
+        uint32_t interfaceIndex = if_nametoindex(current->ifa_name);
+        int family = current->ifa_addr->sa_family;
+        if (family == AF_INET)
+        {
+            // IP Address
+            IpAddressInfo iai;
+            memset(&iai, 0, sizeof(iai));
+            iai.InterfaceIndex = interfaceIndex;
+            sockaddr_in* sain = reinterpret_cast<sockaddr_in*>(current->ifa_addr);
+            memcpy(iai.AddressBytes, &sain->sin_addr.s_addr, sizeof(sain->sin_addr.s_addr));
+            iai.NumAddressBytes = NUM_BYTES_IN_IPV4_ADDRESS;
+
+            // Net Mask
+            IpAddressInfo maskInfo;
+            memset(&maskInfo, 0, sizeof(maskInfo));
+            maskInfo.InterfaceIndex = interfaceIndex;
+            sockaddr_in* mask_sain = reinterpret_cast<sockaddr_in*>(current->ifa_netmask);
+            memcpy(maskInfo.AddressBytes, &mask_sain->sin_addr.s_addr, sizeof(mask_sain->sin_addr.s_addr));
+            maskInfo.NumAddressBytes = NUM_BYTES_IN_IPV4_ADDRESS;
+
+            onIpv4Found(current->ifa_name, &iai, &maskInfo);
+        }
+        else if (family == AF_INET6)
+        {
+            IpAddressInfo iai;
+            memset(&iai, 0, sizeof(iai));
+            iai.InterfaceIndex = interfaceIndex;
+            sockaddr_in6* sain6 = reinterpret_cast<sockaddr_in6*>(current->ifa_addr);
+            memcpy(iai.AddressBytes, sain6->sin6_addr.s6_addr, sizeof(sain6->sin6_addr.s6_addr));
+            iai.NumAddressBytes = NUM_BYTES_IN_IPV6_ADDRESS;
+            uint32_t scopeId = sain6->sin6_scope_id;
+            onIpv6Found(current->ifa_name, &iai, &scopeId);
+        }
+
+        // LINUX : AF_PACKET = 17        
+#if HAVE_AF_PACKET
+        else if (family == AF_PACKET)
+        {
+            LinkLayerAddressInfo lla;
+            memset(&lla, 0, sizeof(lla));
+            lla.InterfaceIndex = interfaceIndex;
+            sockaddr_ll* sall = reinterpret_cast<sockaddr_ll*>(current->ifa_addr);
+            memcpy(&lla.AddressBytes, &sall->sll_addr, sall->sll_halen);
+            lla.NumAddressBytes = sall->sll_halen;
+            lla.HardwareType = sall->sll_hatype;
+            onLinkLayerFound(current->ifa_name, &lla);
+        }
+
+#elif HAVE_AF_LINK
+        // OSX/BSD : AF_LINK = 18
+        else if (family == AF_LINK)
+        {
+            sockaddr_dl* sadl = reinterpret_cast<sockaddr_dl*>(current->ifa_addr);
+            LinkLayerAddressInfo lla;
+            memset(&lla, 0, sizeof(lla));
+            lla.InterfaceIndex = interfaceIndex;
+            memcpy(&lla.AddressBytes, reinterpret_cast<uint8_t*>(LLADDR(sadl)), sadl->sdl_alen);
+            lla.NumAddressBytes = sadl->sdl_alen;
+            lla.HardwareType = sadl->sdl_type;
+            onLinkLayerFound(current->ifa_name, &lla);
+
+            // Do stuff for OSX
+        }
+#endif
```

Comment:
I do have this at the top of the file, but I'm not familiar with what are the best practices for this sort of thing:

````C++
static_assert(HAVE_AF_PACKET || HAVE_AF_LINK, "System must have AF_PACKET or AF_LINK.");
````